plugins/plugin-debugger/che-plugin-debugger-ide/src/main/java/org/eclipse/che/plugin/debugger/ide/debug/BasicActiveFileHandler.java

```diff
@@ -45,42 +45,57 @@ public BasicActiveFileHandler(EditorAgent editorAgent, AppContext appContext) {
     }
 
     /**
-     * Tries to open file in the editor.
+     * Tries to find file and open it.
      * To perform the operation the following sequence of methods invocation are processed:
      *
-     * {@link BasicActiveFileHandler#tryFindFileInProject(Location, AsyncCallback)}
-     * {@link BasicActiveFileHandler#tryFindFileInWorkspace(Location, AsyncCallback)}
-     * {@link BasicActiveFileHandler#trySearchSource(Location, AsyncCallback)}
+     * {@link BasicActiveFileHandler#findInOpenedEditors(Location, AsyncCallback)}
+     * {@link BasicActiveFileHandler#findInProject(Location, AsyncCallback)}
+     * {@link BasicActiveFileHandler#findInWorkspace(Location, AsyncCallback)}
+     * {@link BasicActiveFileHandler#searchSource(Location, AsyncCallback)}
      *
      * @see ActiveFileHandler#openFile(Location, AsyncCallback)
      */
     @Override
     public void openFile(final Location location, final AsyncCallback<VirtualFile> callback) {
-        tryFindFileInProject(location, new AsyncCallback<VirtualFile>() {
+        findInOpenedEditors(location, new AsyncCallback<VirtualFile>() {
+            @Override
+            public void onSuccess(VirtualFile result) {
+                callback.onSuccess(result);
+            }
+
+            @Override
+            public void onFailure(Throwable caught) {
+                findSourceToOpen(location, callback);
+            }
+        });
+    }
+
+    protected void findSourceToOpen(Location location, AsyncCallback<VirtualFile> callback) {
+        findInProject(location, new AsyncCallback<VirtualFile>() {
```

Comment:
Maybe findInProject, findInWorkspace, searchSource should return promise and then this code  can be simplified a bit (less callbacks...)?